In [87]:
# Load my .CSV file into my GitHub repository
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
df = pd.read_csv('https://raw.githubusercontent.com/yoonbs0123/DAV-5400/master/Tidying_Transforming_Data.csv')
df

Unnamed: 0 Unnamed: 1  Los Angeles  Phoenix  San Diego  San Francisco  \
0     ALASKA    on time        497.0    221.0      212.0          503.0   
1        NaN    delayed         62.0     12.0       20.0          102.0   
2        NaN        NaN          NaN      NaN        NaN            NaN   
3     AMWEST    on time        694.0   4840.0      383.0          320.0   
4        NaN    delayed        117.0    415.0       65.0          129.0   

   Seattle  
0   1841.0  
1    305.0  
2      NaN  
3    201.0  
4     61.0

In [88]:
# Rename the columns
df.columns = ['Airline', 'Time', 'Los Angeles', 'Phoenix', 'San Diego', 'San Francisco', 'Seattle']
df

Airline     Time  Los Angeles  Phoenix  San Diego  San Francisco  Seattle
0  ALASKA  on time        497.0    221.0      212.0          503.0   1841.0
1     NaN  delayed         62.0     12.0       20.0          102.0    305.0
2     NaN      NaN          NaN      NaN        NaN            NaN      NaN
3  AMWEST  on time        694.0   4840.0      383.0          320.0    201.0
4     NaN  delayed        117.0    415.0       65.0          129.0     61.0

In [89]:
#drop the rows with all entries as NaN
df2 = df.dropna(how='all',axis=0)
df2

Airline     Time  Los Angeles  Phoenix  San Diego  San Francisco  Seattle
0  ALASKA  on time        497.0    221.0      212.0          503.0   1841.0
1     NaN  delayed         62.0     12.0       20.0          102.0    305.0
3  AMWEST  on time        694.0   4840.0      383.0          320.0    201.0
4     NaN  delayed        117.0    415.0       65.0          129.0     61.0

Using a simple melt transformation with pandas, I was able to convert this untidy data to a “tidy” long format.

In [90]:
df_fill_clean = pd.melt(
    df2,
    id_vars=['Airline', 'Time'],
    var_name='City',
    value_name='Number')
print(df_fill_clean)

   Airline     Time           City  Number
0   ALASKA  on time    Los Angeles   497.0
1      NaN  delayed    Los Angeles    62.0
2   AMWEST  on time    Los Angeles   694.0
3      NaN  delayed    Los Angeles   117.0
4   ALASKA  on time        Phoenix   221.0
5      NaN  delayed        Phoenix    12.0
6   AMWEST  on time        Phoenix  4840.0
7      NaN  delayed        Phoenix   415.0
8   ALASKA  on time      San Diego   212.0
9      NaN  delayed      San Diego    20.0
10  AMWEST  on time      San Diego   383.0
11     NaN  delayed      San Diego    65.0
12  ALASKA  on time  San Francisco   503.0
13     NaN  delayed  San Francisco   102.0
14  AMWEST  on time  San Francisco   320.0
15     NaN  delayed  San Francisco   129.0
16  ALASKA  on time        Seattle  1841.0
17     NaN  delayed        Seattle   305.0
18  AMWEST  on time        Seattle   201.0
19     NaN  delayed        Seattle    61.0


In [91]:
# Correctly fill the NaN items in the 'Airline' column
df_fill_clean['Airline'].fillna(method='ffill', inplace = True)
df_fill_clean

Airline     Time           City  Number
0   ALASKA  on time    Los Angeles   497.0
1   ALASKA  delayed    Los Angeles    62.0
2   AMWEST  on time    Los Angeles   694.0
3   AMWEST  delayed    Los Angeles   117.0
4   ALASKA  on time        Phoenix   221.0
5   ALASKA  delayed        Phoenix    12.0
6   AMWEST  on time        Phoenix  4840.0
7   AMWEST  delayed        Phoenix   415.0
8   ALASKA  on time      San Diego   212.0
9   ALASKA  delayed      San Diego    20.0
10  AMWEST  on time      San Diego   383.0
11  AMWEST  delayed      San Diego    65.0
12  ALASKA  on time  San Francisco   503.0
13  ALASKA  delayed  San Francisco   102.0
14  AMWEST  on time  San Francisco   320.0
15  AMWEST  delayed  San Francisco   129.0
16  ALASKA  on time        Seattle  1841.0
17  ALASKA  delayed        Seattle   305.0
18  AMWEST  on time        Seattle   201.0
19  AMWEST  delayed        Seattle    61.0

In [92]:
# The total number of flights for each airline
at = df_fill_clean.groupby('Airline').sum()
at

Number
Airline        
ALASKA   3775.0
AMWEST   7225.0

In [93]:
# The sum of number, for on time only, to each airline
ao = df_fill_clean[df_fill_clean['Time'] == 'on time'].groupby('Airline')['Number'].sum()
ao

Airline
ALASKA    3274.0
AMWEST    6438.0
Name: Number, dtype: float64

In [94]:
# The sum of number, for delayed only, to each airline
ad = df_fill_clean[df_fill_clean['Time'] == 'delayed'].groupby('Airline')['Number'].sum()
ad

Airline
ALASKA    501.0
AMWEST    787.0
Name: Number, dtype: float64

In [95]:
# find a percentage of on time performance for each airline
ao / (ao+ad) * 100

Airline
ALASKA    86.728477
AMWEST    89.107266
Name: Number, dtype: float64

I found the total number of on time and delayed for each airline seperately. Then, find a percentage of on time performance for each airline by calculating number of flights on time divided by total number of flights for each airline times 100. Even though the values are not much different, I can conclude that AMWEST has a slightly high on time performance compare to that of ALASKA.

In [96]:
# cityontime = sum of number, for on time only, to each city
cityontime = df_fill_clean[df_fill_clean['Time'] == 'on time'].groupby('City')['Number'].sum()
cityontime

City
Los Angeles      1191.0
Phoenix          5061.0
San Diego         595.0
San Francisco     823.0
Seattle          2042.0
Name: Number, dtype: float64

In [97]:
# citytotal = sum of number, for both on time and delayed, to each city
citytotal = df_fill_clean.groupby('City')['Number'].sum()
citytotal

City
Los Angeles      1370.0
Phoenix          5488.0
San Diego         680.0
San Francisco    1054.0
Seattle          2408.0
Name: Number, dtype: float64

In [98]:
# find a percentage of on time performance for each city
cityontime / citytotal * 100

City
Los Angeles      86.934307
Phoenix          92.219388
San Diego        87.500000
San Francisco    78.083491
Seattle          84.800664
Name: Number, dtype: float64

I found the total number of on time only for each city and total number of flights each city seperately. Then, find a percentage of on time performance for each city by calculating cityontime divided by citytotal times 100. Since Phoenix has the highest number, I can conclude that 92.22% of flights in Phoenix are on time. In other words, Phoenix has best on time performance while San Francisco has worst on time performance with only 78.08%. 

Given my 'tidy' long format structure, I would basically mimic the structure of the graphic given in instruction file. To reshape it to wide format from long format, I would use pivot function for that and airlines as index. Therefore, I could get a reshaped dataframe. 